<a href="https://colab.research.google.com/github/gnananawin/Household-power-consumption-Data-analysis/blob/main/Household_Milestone_1_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center><ins>CASE - Household Power Consumption Data: Basic Data Exploration</ins></center>

**<ins>Dataset Description:</ins>**

It measures the power consumption by one household over time, and every entry captures multiple features about the power usage. It has been arranged according to date and time, which captures many of the components of power consumption and the electric properties in the household at any timestamp. It is useful for time-series analysis of energy consumption patterns, short-term forecasting of household energy demand, and studying the impact of different power loads.

**<ins>Description:</ins>**
* Date: The date of the observation.
* Time: The time of the observation.
* Global_active_power: The household's total active power consumption, which represents the real power drawn from the source.
* Global_reactive_power: The household's total reactive power consumption, indicating the power used to maintain the magnetic and electric fields in inductive loads.
* Voltage: The voltage of the household power supply at the time of measurement.
* Global_intensity: The current intensity for the entire household at the time of measurement.
* Sub_metering_1: The active energy consumed by the kitchen appliances, such as dishwasher, oven, and microwave.
* Sub_metering_2: The active energy consumed by the laundry appliances, such as washing machine, tumble dryer, and refrigerator.
* Sub_metering_3: The active energy consumed by the climate control appliances, such as air conditioning and water heater.

**<ins>Dataset Link:</ins>**
https://drive.google.com/file/d/1bvaXJJqNObOCkX-i475BNxpidk024pyx/view

# <center><ins> Problem Statement:</ins></center>
Analyze the **Household Power Consumption** dataset to discover patterns of energy usage over time in a household. This dataset contains minute-level measurements of various power attributes, including active and reactive power consumption, voltage, and intensity, as well as groupings of various appliance types in terms of energy usage.

The objectives of the analysis to be carried out in this report are mainly on key consumption trends, peak usage times, and even factors likely to influence household power demand. Besides, missing values and fluctuations in data have to be managed to analyze correctly. Through this project, the ultimate aim is deriving actionable insights that can be applied in optimizing energy consumption, forecasting demand, and improving household energy efficiency.

# <center><ins> 1.Import Libraries and Basic Data Exploration</ins></center>


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")


In [3]:
!gdown --fuzzy https://drive.google.com/file/d/1bvaXJJqNObOCkX-i475BNxpidk024pyx/view

Downloading...
From (original): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx
From (redirected): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx&confirm=t&uuid=2d747b67-cd87-4ba3-85a7-a2dcfaecee9a
To: /content/household_power_consumption.txt
100% 133M/133M [00:01<00:00, 109MB/s]


In [5]:
df = pd.read_csv('/content/household_power_consumption.txt',sep=";")

In [6]:
df.head()

Date      Time Global_active_power Global_reactive_power  Voltage  \
0  16/12/2006  17:24:00               4.216                 0.418  234.840   
1  16/12/2006  17:25:00               5.360                 0.436  233.630   
2  16/12/2006  17:26:00               5.374                 0.498  233.290   
3  16/12/2006  17:27:00               5.388                 0.502  233.740   
4  16/12/2006  17:28:00               3.666                 0.528  235.680   

  Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0           18.400          0.000          1.000            17.0  
1           23.000          0.000          1.000            16.0  
2           23.000          0.000          2.000            17.0  
3           23.000          0.000          1.000            17.0  
4           15.800          0.000          1.000            17.0

In [7]:
df.shape

(2075259, 9)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075259 entries, 0 to 2075258
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   Date                   object 
 1   Time                   object 
 2   Global_active_power    object 
 3   Global_reactive_power  object 
 4   Voltage                object 
 5   Global_intensity       object 
 6   Sub_metering_1         object 
 7   Sub_metering_2         object 
 8   Sub_metering_3         float64
dtypes: float64(1), object(8)
memory usage: 142.5+ MB


In [9]:
df.tail()

Date      Time Global_active_power Global_reactive_power  \
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
2075254  240.43              4.0            0.0            0.0             0.0  
2075255   240.0              4.0            0.0            0.0             0.0  
2075256  239.82              3.8            0.0            0.0             0.0  
2075257   239.7              3.8            0.0            0.0             0.0  
2075258  239.55              3.8            0.0            0.0             0.0

In [11]:
df.describe()

Sub_metering_3
count    2.049280e+06
mean     6.458447e+00
std      8.437154e+00
min      0.000000e+00
25%      0.000000e+00
50%      1.000000e+00
75%      1.700000e+01
max      3.100000e+01

In [12]:
df.describe(include='object')

Date      Time Global_active_power Global_reactive_power  \
count     2075259   2075259             2075259               2075259   
unique       1442      1440                6534                   896   
top     6/12/2008  17:24:00                   ?                 0.000   
freq         1440      1442               25979                472786   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  
count   2075259          2075259        2075259        2075259  
unique     5168              377            153            145  
top           ?            1.000          0.000          0.000  
freq      25979           169406        1840611        1408274

# <center><ins> Observations:</ins></center>
* With help of df.head() we confirm that the Dataset is correct and contains information about various power consumption metrics.
* To get the dimentions of thedataset we've used df.shape and we got (2075259, 9)
* Here df.tail() shows the last few rows of the dataset.
* The Displayed list of summary statistics for numerical columns are done with df.describe(). When we use (include='object') we get summary of categorical columns like date and time.
* We used df.info() to get a consicise summary of the dataset, including data types and non-null count.

# <center><ins> 2. Finding Unique Values, Null Values, Percentage Of Null Values:</ins></center>

In [13]:
df.nunique()

Date                     1442
Time                     1440
Global_active_power      6534
Global_reactive_power     896
Voltage                  5168
Global_intensity          377
Sub_metering_1            153
Sub_metering_2            145
Sub_metering_3             32
dtype: int64

In [14]:
df.isnull().any()

Date                     False
Time                     False
Global_active_power      False
Global_reactive_power    False
Voltage                  False
Global_intensity         False
Sub_metering_1           False
Sub_metering_2           False
Sub_metering_3            True
dtype: bool

In [15]:
missing_values_count = df.isnull().sum()
print("Missing Values:\n", missing_values_count)

Missing Values:
 Date                         0
Time                         0
Global_active_power          0
Global_reactive_power        0
Voltage                      0
Global_intensity             0
Sub_metering_1               0
Sub_metering_2               0
Sub_metering_3           25979
dtype: int64


In [16]:
missing_percentage = (missing_values_count / len(df)) * 100
print("Missing Values Percentage:\n", missing_percentage.round(2))

Missing Values Percentage:
 Date                     0.00
Time                     0.00
Global_active_power      0.00
Global_reactive_power    0.00
Voltage                  0.00
Global_intensity         0.00
Sub_metering_1           0.00
Sub_metering_2           0.00
Sub_metering_3           1.25
dtype: float64


# <center><ins> Observations:</ins></center>
*  At first, we found the number of unique values present in the dataset which is resulted in a tabular format.
* We can see that Sub_metering_3 is True for any null values which means that parameter of the dataset contains null values.
* Following to that we found the number of missing values(25979) and percentage of missing values(1.25) - here other than sub_metering_3 all parameters have value as 0 as they donot have null values.

# <center><ins> 3.Data Handling For The Missing Values:</ins></center>

In [17]:
df.fillna(0)

Date      Time Global_active_power Global_reactive_power  \
0        16/12/2006  17:24:00               4.216                 0.418   
1        16/12/2006  17:25:00               5.360                 0.436   
2        16/12/2006  17:26:00               5.374                 0.498   
3        16/12/2006  17:27:00               5.388                 0.502   
4        16/12/2006  17:28:00               3.666                 0.528   
...             ...       ...                 ...                   ...   
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
2075254   240.43              4.0            0.0            0.0   
2075255    240.0              4.0            0.0            0.0   
2075256   239.82              3.8            0.0            0.0   
2075257    239.7              3.8            0.0            0.0   
2075258   239.55              3.8            0.0            0.0   

         Sub_metering_3  
0                  17.0  
1                  16.0  
2                  17.0  
3                  17.0  
4                  17.0  
...                 ...  
2075254             0.0  
2075255             0.0  
2075256             0.0  
2075257             0.0  
2075258             0.0  

[2075259 rows x 9 columns]

In [18]:
df.fillna(df['Sub_metering_3'].mean())

Date      Time Global_active_power Global_reactive_power  \
0        16/12/2006  17:24:00               4.216                 0.418   
1        16/12/2006  17:25:00               5.360                 0.436   
2        16/12/2006  17:26:00               5.374                 0.498   
3        16/12/2006  17:27:00               5.388                 0.502   
4        16/12/2006  17:28:00               3.666                 0.528   
...             ...       ...                 ...                   ...   
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
2075254   240.43              4.0            0.0            0.0   
2075255    240.0              4.0            0.0            0.0   
2075256   239.82              3.8            0.0            0.0   
2075257    239.7              3.8            0.0            0.0   
2075258   239.55              3.8            0.0            0.0   

         Sub_metering_3  
0                  17.0  
1                  16.0  
2                  17.0  
3                  17.0  
4                  17.0  
...                 ...  
2075254             0.0  
2075255             0.0  
2075256             0.0  
2075257             0.0  
2075258             0.0  

[2075259 rows x 9 columns]

In [19]:
df.fillna(df['Sub_metering_3'].median())

Date      Time Global_active_power Global_reactive_power  \
0        16/12/2006  17:24:00               4.216                 0.418   
1        16/12/2006  17:25:00               5.360                 0.436   
2        16/12/2006  17:26:00               5.374                 0.498   
3        16/12/2006  17:27:00               5.388                 0.502   
4        16/12/2006  17:28:00               3.666                 0.528   
...             ...       ...                 ...                   ...   
2075254  26/11/2010  20:58:00               0.946                   0.0   
2075255  26/11/2010  20:59:00               0.944                   0.0   
2075256  26/11/2010  21:00:00               0.938                   0.0   
2075257  26/11/2010  21:01:00               0.934                   0.0   
2075258  26/11/2010  21:02:00               0.932                   0.0   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
2075254   240.43              4.0            0.0            0.0   
2075255    240.0              4.0            0.0            0.0   
2075256   239.82              3.8            0.0            0.0   
2075257    239.7              3.8            0.0            0.0   
2075258   239.55              3.8            0.0            0.0   

         Sub_metering_3  
0                  17.0  
1                  16.0  
2                  17.0  
3                  17.0  
4                  17.0  
...                 ...  
2075254             0.0  
2075255             0.0  
2075256             0.0  
2075257             0.0  
2075258             0.0  

[2075259 rows x 9 columns]

In [20]:
df.dropna().shape

(2049280, 9)

# <center><ins> Observations:</ins></center>
The dataset contains many null values for a particular parameter. We have many methods to fill the null values and some of the methods are followed here to handle null values for the given dataset.
* Firstly, we used df.fillna(0) which is not used most of the time but it can be only used when we need to fill the null values with 0. After the following is done we can see the resultant dataset.
* Secondly, we used df.fillna(df['Sub_metering_3'].mean()) which will first find the mean value of that particular parameter and then it replaces the null values with that mean value.
* Next to that we used df.fillna(df['Sub_metering_3'].median()) which is similar to the second method but here we first find the median value of the parameter and rearrange the null values with the median value.
* Again, We just compute the dimention of the whole dataset for our reference